File: spanning_tree.py
from operator import xor
from edge import Edge

class SpanningTree:

    def __init__(self, graphSize):

        self.edges = []
        self.vertex_degrees = []
        for i in range(0, graphSize):
            self.vertex_degrees.append(0)

    def edgeList(self):
        return self.edges

    def vertexDegrees(self):
        return self.vertex_degrees

    def addEdge(self, edge):
        self.vertex_degrees[edge.vertex1] += 1
        self.vertex_degrees[edge.vertex2] += 1
        self.edges.append(edge)

    def removeLastEdge(self):
        removed = self.edges.pop()
        self.vertex_degrees[removed.vertex1] -= 1
        self.vertex_degrees[removed.vertex2] -= 1

    def edgeCount(self):
        return len(self.edges)

    def maxDegree(self):
        return max(self.vertex_degrees)

    def maxDegreeWith(self, edge):
        max = self.vertex_degrees[0]

        for i in range(0, len(self.vertex_degrees)):
            degree = self.vertex_degrees[i]
            if (edge.vertex1 == i) or (edge.vertex2 == i):
                degree += 1
            if degree > max:
                max = degree

        return max

    def __str__(self):
        result = ", ".join(str(edge) for edge in self.edges)
        result += "\ndegree: {0}".format(self.maxDegree())
        return result

File: serial/KGM-JS/SpanningTree.cpp
/* 
 * File:   SpanningTree.cpp
 */

#include "SpanningTree.h"

using namespace std;

SpanningTree::SpanningTree(int vertexCount) {
	this->vertexCount = vertexCount;
	edges = new vector<Edge>();
	vertexDegrees = new int[vertexCount];
	std::fill_n(vertexDegrees, vertexCount, 0);
}

SpanningTree::SpanningTree(const SpanningTree& orig) {
	this->vertexCount = orig.vertexCount;
	edges = new vector<Edge>(*orig.edges);
	vertexDegrees = new int[vertexCount];
	std::copy(orig.vertexDegrees, orig.vertexDegrees + vertexCount, vertexDegrees);	
}

SpanningTree::~SpanningTree() {
	delete edges;
	delete [] vertexDegrees;
}

void SpanningTree::addEdge(Edge edge) {
	vertexDegrees[edge.vertex1]++;
	vertexDegrees[edge.vertex2]++;
	edges->push_back(edge);
}

void SpanningTree::removeLastEdge() {
	Edge removed = edges->back();
	vertexDegrees[removed.vertex1]--;
	vertexDegrees[removed.vertex2]--;
	edges->pop_back();
}

int SpanningTree::edgeCount() const {
	return edges->size();
}

int SpanningTree::evaluate() const {
	int max = vertexDegrees[0];
	for (int i = 0; i < vertexCount; i++) {
		if (vertexDegrees[i] > max) {
			max = vertexDegrees[i];
		}
	}
	return max;
}

int SpanningTree::evaluate(Edge current) const {
	int max = vertexDegrees[0];
	for (int i = 0; i < vertexCount; i++) {
		int vertexDeg = vertexDegrees[i]; 
		if (current.vertex1 == i || current.vertex2 == i) {
			vertexDeg++;
		}
		if (vertexDeg > max) {
			max = vertexDeg;
		}
	}
	return max;	
}

int * SpanningTree::getVertexDegrees() const {
	return vertexDegrees;
}

vector<Edge> * SpanningTree::getEdges() const {
	return edges;
}

std::ostream & operator <<(std::ostream & os, const SpanningTree & tree) {
	for (int i = 0; i < tree.edges->size(); i++) {
		if (i != 0) {
			os << ", ";
		}
		Edge & edge = (*tree.edges)[i];
		os << edge;
	}
	os << endl << "degree: " << tree.evaluate() << endl;
	return os;
}

void SpanningTree::printVertexDegrees() {
	cout << "| "; 
	for (int i = 0; i < vertexCount; i++) {
		cout << vertexDegrees[i] << " | ";
	}
	cout << endl;
}